### PR TITLE
Fix sidecar refinement / MOUNT volume reuse

### PR DESCRIPTION
**Pull request by gabrielhartmann**
### Sidecar Resource Refinement Problem
You need to do 3 things to encounter the bug fixed here.
1. Have a volume at the pod level
1. Have a sidecar task on that pod
1. Consume pre-reserved resources for that pod

`edge-lb` does all three of those things and so can't use `0.40.X` versions of the SDK.  This should fix that.  I'll confirm with them before release of `0.40.2`.

### Fix
This change can be a little bit hard to follow.  There were a couple things that needed to happen.
1. Use the `ResourceBuilder` class to build executor volumes so you get the reservation refinement stuff.
1. Pass around the `sourceRoot` field so we can build `MOUNT` volumes for Executor reuse.

### MOUNT volume problem
We have moved to a consistent construction pattern for protobufs.  That is we don't pull them out of the StateStore and reuse them _ever_.  This had problems for mount volumes in particular, because they're atomic.  The `VolumeSpec` says use `50 GB`, but the MOUNT was `100 GB`.  When reusing an Executor volume we need to indicate `100 GB` not fifty.  That information isn't in the spec.

### Fix
When matching resources and specs, grab the size info and put it in the spec.

### Note
~~The custom executor is only supported on 1.9 and previous.  1.9 and previous clusters do not support reservation refinement.  That's why the custom executor code path remains unmodified.~~
This has been manually tested on 1.9 and 1.10.  It's been tested with DSE as well.

We have to solve the generating clusters with MOUNT volumes problem to get integration tests going again.  This PR isn't going to do that.

### CODE REVIEW

**Comment by nickbp** [comment]
Still WIP?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/VolumeEvaluationStage.java [hunk 1]
```diff
@@ -58,10 +101,13 @@ public EvaluationOutcome evaluate(MesosResourcePool mesosResourcePool, PodInfoBu
...
 
+            String sourceRoot = null;
+
             Resource volume = PodInfoBuilder.getExistingExecutorVolume(
                     volumeSpec,
-                    resourceId.get(),
-                    persistenceId.get(),
+                    resourceId,
+                    persistenceId,
+                    Optional.ofNullable(sourceRoot),
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/VolumeEvaluationStage.java]: Always null?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/ResourceUtils.java [hunk 2]
```diff
@@ -115,10 +115,18 @@ public static boolean hasResourceId(Resource resource) {
     }
 
     public static Optional<String> getSourceRoot(Resource resource) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/ResourceUtils.java]: Should VolumeEvaluationStage be calling this? Also, can't hurt to check its current callers (there could very well not be any)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/TaskUtils.java [hunk 3]
```diff
@@ -531,7 +531,9 @@ public static boolean needsRecovery(TaskSpec taskSpec, TaskStatus taskStatus) {
         // Tasks with a goal state of finished should never leave the purview of their original
         // plan, so they are not the responsibility of recovery.  Recovery only applies to Tasks
         // which reached their goal state of RUNNING and then later failed.
-        if (taskSpec.getGoal().equals(GoalState.ONCE) || taskSpec.getGoal().equals(GoalState.FINISH)) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/TaskUtils.java]: Could have a switch statement that explicitly lists the goal states. IDEs/linters can then catch if a goal state is created and then isn't added here.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ExecutorResourceMapper.java [hunk 4]
```diff
@@ -171,23 +147,40 @@ public ExecutorResourceMapper(
...
+                resourceLabels.getSourceRoot());
     }
 
     private OfferEvaluationStage newCreateEvaluationStage(ResourceSpec resourceSpec) {
-        return toEvaluationStage(resourceSpec, Optional.empty(), Optional.empty());
+        return toEvaluationStageForNew(resourceSpec);
     }
 
-    private OfferEvaluationStage toEvaluationStage(
+    private OfferEvaluationStage toEvaluationStageForExisting(
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ExecutorResourceMapper.java]: Given that this is only called by newUpdateEvaluationStage(), its content could be folded into there.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ExecutorResourceMapper.java [hunk 5]
```diff
@@ -171,23 +147,40 @@ public ExecutorResourceMapper(
...
+                    resourceId,
+                    persistenceId,
+                    sourceRoot,
+                    useDefaultExecutor);
         } else {
             return new ResourceEvaluationStage(resourceSpec, resourceId, null);
         }
     }
+
+    private OfferEvaluationStage toEvaluationStageForNew(ResourceSpec resourceSpec) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ExecutorResourceMapper.java]: Given that this is only called by newCreateEvaluationStage(), its content could be folded into there.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/VolumeEvaluationStage.java (hunk 1 above)
Comment (gabrielhartmann) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/VolumeEvaluationStage.java]: No, that was part of the WIP version.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/ResourceUtils.java (hunk 2 above)
Comment (gabrielhartmann) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/ResourceUtils.java]: WIP artifact.   Should be called appropriately now.

**Comment by gabrielhartmann** [comment]
@nickbp: Please have another look while I write a ServiceTest for MOUNT volumes.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/VolumeEvaluationStage.java [hunk 6]
```diff
@@ -25,24 +25,61 @@
...
         this.persistenceId = persistenceId;
+        this.sourceRoot = sourceRoot;
         this.useDefaultExecutor = useDefaultExecutor;
     }
 
     private boolean createsVolume() {
         return !persistenceId.isPresent();
     }
 
+    private static boolean isMountVolume(MesosResource mesosResource) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/VolumeEvaluationStage.java]: Duplicate of `ResourceUtils.isMountVolume()`. Might not even be used? Diff doesn't show anyone calling into this.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TaskResourceMapper.java [hunk 7]
```diff
@@ -17,7 +16,7 @@
  * Handles cross-referencing a preexisting {@link Protos.TaskInfo}'s current {@link Protos.Resource}s against a set
  * of expected {@link ResourceSpec}s for that task.
  */
-class TaskResourceMapper {
+public class TaskResourceMapper {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TaskResourceMapper.java]: Is this actually used outside of this package now? Didn't see any external usage in this change. But feel free to ignore if external usage is coming soon.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ResourceLabels.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    public ResourceLabels(ResourceSpec resourceSpec, String resourceId) {
+        this(resourceSpec, resourceSpec, resourceId);
+    }
+
+    public ResourceLabels(ResourceSpec original, ResourceSpec updated, String resourceId) {
+        this(original, updated, resourceId, Optional.empty(), Optional.empty());
+    }
+
+    public ResourceLabels(
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ResourceLabels.java]: IMO could stand to have some javadocs around `original` vs `updated`.

Comment (gabrielhartmann) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ResourceLabels.java]: yep.  I'll add some.
